src/AsyncEventSource.cpp

```diff
@@ -25,114 +25,97 @@
 
 using namespace asyncsrv;
 
-static String generateEventMessage(const char* message, const char* event, uint32_t id, uint32_t reconnect) {
-  String ev;
+static std::string generateEventMessage(const char* message, const char* event, uint32_t id, uint32_t reconnect) {
+  //String ev;
+  std::string str;
+  size_t len{0};
+  if (message)
+    len += strlen(message);
+
+  if (event)
+    len += strlen(event);
+
+  len += 42;    // give it some overhead
+
+  str.reserve(len);
 
   if (reconnect) {
-    ev += T_retry_;
-    ev += reconnect;
-    ev += T_rn;
+    str += T_retry_;
+    str += reconnect;
+    str += 0xa; // '\n'
   }
 
   if (id) {
-    ev += T_id__;
-    ev += id;
-    ev += T_rn;
+    str += T_id__;
+    str += id;
+    str += 0xa; // '\n'
   }
 
   if (event != NULL) {
-    ev += T_event_;
-    ev += event;
-    ev += T_rn;
+    str += T_event_;
+    str += event;
+    str += 0xa; // '\n'
```

Comment:
- call `s.reserve()` to pre-allocate the right heap amont and avoir inner buffer copies: the size can be computed upfront since thsi is just a concatenation
- using `s.concat(...)` would avoid the useless re-assignement operation everywhere
- even better: since this method is called often and could be called with a lot of data, its signature could be changd to: `static void generateEventMessage(String& buffer, const char* message, const char* event, uint32_t id, uint32_t reconnect)`, then the buffer is reserved then filled.

Reply:
yes, I'm doing `str.reserve(len);` just above this lines,  `+=` is equivalent to `.append()` and does not relocate if buffer side is wide enough.

`String` is not acceptable here, `AsyncEventSourceMessage` container is `std::shared_ptr<std::string>`, because `std::string` has proper move constructor and Arduino's `String` does not.
I create `std::string` object and fill it with SSE's formatted message from supplied `const char*`, then the object is used to move into `AsyncEventSourceMessage` constructor and create a shared pointer for the message emplaced directly in the queue, no data is copied or relocated, that is the whole idea.
Not sure I would be able to achieve the same with Arduino's `String`.

With `generateEventMessage(String& buffer, const char* message...` would require to pre calculate and make the `String` each time outside the function while I do the preallocation inside the function despite the arguments passed.

Reply:
> yes, I'm doing str.reserve(len); just above this lines,
missed it, sorry!

> because std::string has proper move constructor and Arduino's String does not.
yes, that's right.

I don't want to do any debate about which one is best, but I just wonder. is std::string also allocate on head like String ? Since these messages could be important in size, allocation should not be done within the stack. My assumption was that std:string was allocating on stack (but I am not familiar enough).

Reply:
Ok I saw the answer here with the move semantic:

https://stackoverflow.com/questions/3976935/best-way-to-return-an-stdstring-that-local-to-a-function

The Arduino String implementation does not have that baked in indeed. I definitely need to know more about that!

Reply:
yes, it should optimize to move construct. As for `std::string` allocation - it is compiller implementation specific and if espressif has not optimized it on their own (which I doubt honestly :) ) it should use some kind of malloc() wrapper under the hood and and up in the heap. So should be no problem with stack mem.
I totally agree to stick to `String`s for the user-side API but if internal implementations could benefit from `std` containers then why not use it? :)